code/modules/maps/template_types/random_exoplanet/fauna_generator.dm

```diff
@@ -0,0 +1,273 @@
+	return new /datum/generated_fauna_template(spawn_type, atmos_temp, min_gas, max_gas)
+
+/datum/fauna_generator/proc/try_respawn_from_queue()
+	. = LAZYACCESS(respawn_queue, 1)
+	if(.)
+		respawn_queue = respawn_queue.Copy(2) //Clear first entry
+
+/datum/fauna_generator/proc/pick_fauna_to_respawn()
+	. = LAZYACCESS(respawn_queue & fauna_types, 1)
+	if(.)
+		respawn_queue = respawn_queue.Copy(2) //Clear first entry
+	else
+		//Grab from our fauna list if we don't have anything set to respawn
+		. = pickweight(fauna_types)
+
+/datum/fauna_generator/proc/pick_megafauna_to_respawn()
+	. = LAZYACCESS(respawn_queue & megafauna_types, 1)
+	if(.)
+		respawn_queue = respawn_queue.Copy(2) //Clear first entry
+	else
+		//Grab from our fauna list if we don't have anything set to respawn
+		. = pickweight(megafauna_types)
+
+/datum/fauna_generator/proc/try_spawn_fauna(var/turf/T, var/force = FALSE)
+	if(!LAZYLEN(fauna_types))
+		log_warning("Tried spawning fauna at ([T.x], [T.y], [T.z]) but the fauna_types list is empty!")
+		return
+	if(!force && LAZYLEN(live_fauna) >= max_fauna_alive)
+		return
+	var/spawning = pick_fauna_to_respawn()
+	var/datum/generated_fauna_template/Tmpl = fauna_templates[spawning]
+	var/datum/planetoid_data/P              = SSmapping.planetoid_data_by_z[T.z]
+	. = Tmpl.try_spawn(T, P)
+
+	//Add to tracking
+	register_fauna(.)
+
+/datum/fauna_generator/proc/try_spawn_megafauna(var/turf/T, var/force = FALSE)
+	if(!LAZYLEN(megafauna_types))
+		log_warning("Tried spawning megafauna at ([T.x], [T.y], [T.z]) but the megafauna_types list is empty!")
+		return
+	if(!force && LAZYLEN(live_megafauna) >= max_megafauna_alive)
+		return
+	var/spawning = pick_megafauna_to_respawn()
+	var/datum/generated_fauna_template/Tmpl = megafauna_templates[spawning]
+	var/datum/planetoid_data/P              = SSmapping.planetoid_data_by_z[T.z]
+	. = Tmpl.try_spawn(T, P)
+
+	//Add to tracking
+	register_fauna(.)
+
+//#TODO: move to subsystem/manager
+/datum/fauna_generator/Process()
+	if(time_last_repop < (time_last_repop + repopulation_interval))
+		return //Wait a bit before repopulating
+	if(LAZYLEN(respawn_queue) > 0)
+		handle_repopulation()
+
+/datum/fauna_generator/proc/handle_repopulation()
+	var/datum/level_data/LD = SSmapping.levels_by_id[level_data_id]
+	var/area/spawn_area = LD.get_base_area_instance()
+
+	for(var/i = 1 to round(repopulate_megafauna_threshold - length(live_megafauna)))
+		if(prob(90))
+			continue
+		var/turf/simulated/T = pick_area_turf(spawn_area, list(/proc/not_turf_contains_dense_objects))
+		try_spawn_megafauna(T)
+
+	for(var/i = 1 to round(repopulate_fauna_threshold - length(live_fauna)))
+		if(prob(90))
+			continue
+		var/turf/simulated/T = pick_area_turf(spawn_area, list(/proc/not_turf_contains_dense_objects))
+		try_spawn_fauna(T)
+
+	time_last_repop = REALTIMEOFDAY
+	//Half the repop threshold each times we empty the respawn list
+	if(LAZYLEN(respawn_queue) <= 0)
```

Comment:
Wouldn't this cause threshold to drop p fast even if not a lot of animals are killed? If just 1 animal dies per 5 minutes it'll be depopulated in 20 minutes (right away in casae of megafauna)

Reply:
I don't really know, that was the old fauna code I ported into the fauna generator. So I assumed it worked.

Reply:
Well, if that's old code can be figured out later I guess